File: android/src/main/kotlin/ai/amani/flutter_amanisdk/modules/DocumentCapture.kt
package ai.amani.flutter_amanisdk.modules

import ai.amani.flutter_amanisdk.R
import ai.amani.sdk.Amani
import ai.amani.sdk.interfaces.IUploadCallBack
import ai.amani.sdk.modules.document.DocBuilder
import ai.amani.sdk.modules.document.FileWithType
import ai.amani.sdk.modules.document.interfaces.IDocumentCallBack
import android.app.Activity
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream

class DocumentCapture: Module {

    private var docType: String? = null;
    private var frag: Fragment? = null;
    private var closeButton: Button? = null;
    private var filesToUpload: ArrayList<FileWithType>? = null

    companion object {
        val instance = DocumentCapture()
    }

    // stepID is DOCUMENT COUNT
    override fun start(stepID: Int, activity: Activity, result: MethodChannel.Result) {
        if (docType == null) {
            result.error("30013", "You must call setType before calling start method", null)
            return
        }

        (activity as FragmentActivity)

        val id = 0x001115
        val context = activity.applicationContext
        val viewParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        val container = FrameLayout(context)

        container.id = id
        activity.addContentView(container, viewParams)

        frag = Amani.sharedInstance().Document().start(docType, null, container, object: IDocumentCallBack {
            override fun cb(listOfDocumentAbsolutePath: ArrayList<String>?, isSuccess: Boolean) {
                if (listOfDocumentAbsolutePath?.last() == null) {
                    result.error("30013", "No images are returned from DocumentCapture", null)
                }

                val bitmap: Bitmap = BitmapFactory.decodeFile(listOfDocumentAbsolutePath?.last())
                val stream = ByteArrayOutputStream()
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, stream);
                result.success(stream.toByteArray())

                // Result returned close the view!
                activity.supportFragmentManager.beginTransaction()
                    .remove(frag!!)
                    .commit()

                activity.runOnUiThread {
                    closeButton!!.visibility = View.GONE
                }
            }
        })

        closeButton = container.setupBackButton(R.drawable.baseline_close_24, onClick = {
            activity.supportFragmentManager.beginTransaction().remove(frag!!).commit()
        })

        activity.supportFragmentManager.beginTransaction()
            .replace(id, frag!!)
            .commit()
    }

    fun backPressHandle(activity: Activity, result: MethodChannel.Result) {
        if (frag == null) {
            result.error(
                "30001",
                "You must call this function while the" +
                        "module is running", "You can ignore this message and return true" +
                        "from onWillPop()"
            )
        } else {
            activity.runOnUiThread {
                frag?.let {
                    closeButton!!.visibility = View.GONE
                    it.parentFragmentManager.beginTransaction().remove(frag!!).commit()
                    frag = null
                    // This blocks the flutters back press action.
                    result.success(false)
                }
            }
        }
    }

    //                                   data, dataType
    //                                   ByteArray, string
   fun setFiles(files: List<Map<String, Any>>) {
        filesToUpload = files.map {
            val data = it.getValue("data") as ByteArray
            val type = it.getValue("dataType") as String
            FileWithType(data, type)
        } as ArrayList<FileWithType>
    }

    override fun upload(activity: Activity, result: MethodChannel.Result) {

        if (docType == null) {
            result.error("30013", "You must call setType before calling start method", null)
            return
        }

        (activity as FragmentActivity)
        if (filesToUpload == null) {
            Amani.sharedInstance().Document().upload(activity, docType!!, object : IUploadCallBack {
                override fun cb(uploadState: Boolean) {
                    result.success(uploadState)
                }
            })
        } else {
            Amani.sharedInstance().Document().upload(activity, docType!!, filesToUpload!!) {
                result.success(it)
            }
        }
    }

    override fun setType(type: String?, result: MethodChannel.Result) {
        this.docType = type
        result.success(null)
    }


}

File: ios/Classes/modules/DocumentCapture.swift
//
//  DocumentCapture.swift
//  flutter_amanisdk
//
//  Created by Deniz Can on 27.02.2024.
//

import AmaniSDK
import Flutter
import Foundation

class DocumentCapture {
  private let module = Amani.sharedInstance.document()
  private var sdkView: SDKView!

  func start(documentCount: Int? = 1, result: @escaping FlutterResult) {
    let vc = UIApplication.shared.windows.last?.rootViewController

    do {
      if let moduleView = try module.start(stepId: documentCount!, completion: { image in
        if let data = image.pngData() {
          result(FlutterStandardTypedData(bytes: data))
        }
        DispatchQueue.main.async {
          self.sdkView.removeFromSuperview()
        }
      }) {
        sdkView = SDKView(sdkView: moduleView)
        sdkView.start(on: vc!)
        sdkView.setupBackButton(on: moduleView)
      }
    } catch {
      result(FlutterError(code: "30007", message: error.localizedDescription, details: nil))
    }
  }

  func setType(type: String, result: @escaping FlutterResult) {
    module.setType(type: type)
    result(nil)
  }

  func upload(files: [FileWithType]?, result: @escaping FlutterResult) {
    if let files = files {
      module.upload(location: nil, files: files) { uploadRes, _ in
        result(uploadRes)
      }
    } else {
      module.upload(completion: { uploadRes in
        result(uploadRes)
      })
    }
  }
}
